### DISCUSSION TITLE
What is the scope of a form?

### DISCUSSION THREAD

**Discussion by msageryd** [react-hook-form/react-hook-form#7546] [Q&A] [answered]
This seems like a very nice for lib. I'm investigating to see if I should migrate from Final-Form.

useForm will get me access to form functions from anywhere. When does a form "start" and "stop"? In React-Native I don't even have a `<form>` tag to enclose the form. I have many forms in my app. What's stopping all those form "being" the same form when they all get the same form functions from useForm.

Example:
I register a field "firstName" in form "user" with `register` from useForm.
How log will this field stay registered?
When I enter form 'company' and register field "companyName", will firstName still be registered?

There is no wrapping component which can provide the context for the form, so I'm a bit confused.

**Comment by Moshyfawn** [comment, marked as answer]
That's a very broad question right here 🤪
The basic concept of RHF is you **register** your input **with a particular form instance** by calling [`register`](https://react-hook-form.com/api/useform/register) on it from the current [`useForm` hook](https://react-hook-form.com/api/useform). In other words, your form starts and ends with all the registered inputs to a particular `useForm`, not only what's wrapped in a `<form />` tag.
> "How log will this field stay registered?"

This depends on your [`useForm` configuration](https://react-hook-form.com/api/useform/) or the use of the [`unregister` method _(more advanced)_](https://react-hook-form.com/api/useform/unregister). Simply put, by default, an input value will be retained when an input is unmounted. But there's a `useForm` prop `shouldUnregister`, that allows you to tell the form to get rid of _(unregister)_ the unmounted input and it's value. Basically, close to what native form's behavior is like.
All form inputs will unregister when the component with `useForm` hook unmounts.

Like I said, it's very broad question and I don't want you to get lost in all those words. So, feel free to ask further questions.
Check out [this codesandbox](https://codesandbox.io/s/use-form-example-lh5sg?file=/src/App.tsx) with the RHF devtools on the right side. Play around with the UI, see what's displayed in the devtools when switch forms / hide inputs. Code-wise, there's **forms folder** with company and user forms code. Notice how both forms have their `useForm` hook and it's `register` method.

**Reply by msageryd** [reply]
Thanks.
I should explain my use case..

My forms are completely declarative. All fields, validation rules, etc is provided in a JSON from my REST API. Due to some complicated cross-referenced condition-checking I need to register all fields, even if only some are shown from start. Fields can show/hide due to rules in my conditions-engine.

I don't want my fields to automatically unregsiter when they are unMounted. So, if I understand you correctly, I should take resposibility of the lifecycle of my forms by:
1. register fields upon form-start
2. unregister fields upon form-end (i.e. when the form screen is closed)

Are there any more cleanup I need to do, besides unregistering all the fields, before I can reuse `useForm` for another form? I.e. will isValid, etc be reset when all fields are unregistered?

**Reply by msageryd** [reply]
Oh, I see now that you wrote "All form inputs will unregister when the component with useForm hook unmounts."

This seems to imply that the scope of a form is:
- start = when useForm is called
- end = when useForm-component is unmounted

In that case I don't need to bother with unregister.

**Reply by Moshyfawn** [reply]
> This seems to imply that the scope of a form is:
> 
> - start = when useForm is called
> - end = when useForm-component is unmounted

That's correct! Sorry, I could've phrased it better.

> In that case I don't need to bother with unregister.

`unregister` method comes into play in case there's a need to unregister an input without unmounting it _(i.e. visible disabled input, the value of which shouldn't be validated, taken into account on submit)_. Closing the form screen _(form component unmount)_ should handle all the necessary cleanup.

In case you want to reuse the current screen to start a new form, you can use [the `reset` form method](https://react-hook-form.com/api/useform/reset/#main). Take note of this rule:
> When you are subscribed to formState, it's important to decouple reset with handleSubmit.
> Because onSubmit callback is async and includes its validation when reset inside the callback it will intercept the result of formState update. This will be problematic when you subscribed to the formState.

**Reply by msageryd** [reply]
Thank you for your great clarifications.
This seems to be a very clean and user friendly library. I've only looked at it for a couple of ours and I'm already eager to replace Final-Form with react-hook-form.
